cpp/libcudf_kafka/src/kafka_consumer.cpp

```diff
@@ -112,6 +133,101 @@ void kafka_consumer::consume_to_buffer()
   }
 }
 
+std::map<std::string, std::string> kafka_consumer::current_configs()
+{
+  std::map<std::string, std::string> configs;
+  std::list<std::string> *dump = kafka_conf->dump();
+  std::string key;
+  std::string val;
+  for (std::list<std::string>::iterator it = dump->begin(); it != dump->end();) {
+    key = (*it);
+    it++;
+    val = (*it);
+    it++;
+    configs.insert(std::pair<std::string, std::string>{key, val});
+  }
+  return configs;
+}
+
+int64_t kafka_consumer::get_committed_offset(std::string const &topic, int partition)
+{
+  std::vector<RdKafka::TopicPartition *> toppar_list;
+  toppar_list.push_back(RdKafka::TopicPartition::create(topic, partition));
+
+  // Query Kafka to populate the TopicPartitions with the desired offsets
+  RdKafka::ErrorCode err = consumer->committed(toppar_list, default_timeout);
+
+  int64_t offset = toppar_list[0]->offset();
+  return offset > 0 ? offset : 0;
+}
+
+std::map<std::string, int64_t> kafka_consumer::get_watermark_offset(std::string const &topic,
+                                                                    int partition,
+                                                                    int timeout,
+                                                                    bool cached)
+{
+  int64_t low;
+  int64_t high;
+  std::map<std::string, int64_t> results;
+  RdKafka::ErrorCode err;
+
+  if (cached == true) {
+    err = consumer->get_watermark_offsets(topic, partition, &low, &high);
+  } else {
+    err = consumer->query_watermark_offsets(topic, partition, &low, &high, timeout);
+  }
+
+  if (err != RdKafka::ErrorCode::ERR_NO_ERROR) {
+    if (err == RdKafka::ErrorCode::ERR__PARTITION_EOF) {
+      results.insert(std::pair<std::string, int64_t>("low", low));
+      results.insert(std::pair<std::string, int64_t>("high", high));
+    } else {
+      CUDF_FAIL("Error retrieving Kafka watermark offset from broker");
+    }
+  } else {
+    results.insert(std::pair<std::string, int64_t>("low", low));
+    results.insert(std::pair<std::string, int64_t>("high", high));
+  }
+
+  return results;
+}
+
+bool kafka_consumer::commit_offset(std::string const &topic, int partition, int64_t offset)
+{
+  std::vector<RdKafka::TopicPartition *> partitions_;
+  RdKafka::TopicPartition *toppar = RdKafka::TopicPartition::create(topic, partition, offset);
+  if (toppar != NULL) {
+    toppar->set_offset(offset);
+    partitions_.push_back(toppar);
+    RdKafka::ErrorCode err = consumer->commitSync(partitions_);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool kafka_consumer::unsubscribe()
+{
+  RdKafka::ErrorCode err = consumer.get()->unassign();
+  if (err != RdKafka::ERR_NO_ERROR) {
+    printf(
+      "Timeout occurred while unsubscribing from Kafka Consumer "
+      "assignments.\n");
+    return false;
+  } else {
+    return true;
+  }
+}
+
+bool kafka_consumer::close(int timeout)
+{
+  RdKafka::ErrorCode err = consumer.get()->close();
+  delete consumer.get();
```

Comment:
@harrism Just want to make sure I'm not missing something but this should be `consumer.reset(nullptr)` and not `consumer->reset(nullptr)` which would attempt to dereference and call the function `reset` on the `consumer` object right?

Reply:
also since both `consumer` and `kafka_conf` are C implementations would I not need to explicitly call `delete` on them? This is from the Kafka docs [here](https://docs.confluent.io/2.0.0/clients/librdkafka/classRdKafka_1_1KafkaConsumer.html#a5c78a721aa91f3be9903f09ddf084644) at the bottom of the page, sorry can get an embedded link for it but it says ... `The consumer object must later be freed with delete`

Reply:
https://en.cppreference.com/w/cpp/memory/unique_ptr/reset

> 1) Given current_ptr, the pointer that was managed by *this, performs the following actions, in this order:
Saves a copy of the current pointer old_ptr = current_ptr
Overwrites the current pointer with the argument current_ptr = ptr
If the old pointer was non-empty, deletes the previously managed object if(old_ptr) get_deleter()(old_ptr).

Now, `get_deleter()` is going to return `std::default_delete<T>` (https://en.cppreference.com/w/cpp/memory/default_delete)

Which says: 
> 1) The non-specialized default_delete uses delete to deallocate memory for a single object.

So this means it will do the equivalent of the code you had before, except that now the `unique_ptr` will not have a dangling pointer to the thing you deleted.